geopyspark/geotrellis/key_conversion.py

```diff
@@ -0,0 +1,119 @@
+        crs (str or int): Used only when `layout` is :class:`geopyspark.geotrellis.GlobalLayout`.  Target CRS of reprojection.
+            Either EPSG code, well-known name, or a PROJ.4 string
+        extent (:class:`geopyspark.geotrellis.Extent`): Used only for ``LocalLayout``s.  The area of interest.
+        cellsize (tup of (float, float)): Used only for ``LocalLayout``s.  The (width, height) in extent units of a pixel.
+            Cannot be specified simultaneously with ``dimensions``.
+        dimensions (tup of (int, int)): Used only for ``LocalLayout``s.  The number of (columns, rows) of pixels over the
+            entire extent.  Cannot be specified simultaneously with ``cellsize``.
+    """
+
+    def __init__(self, layout, crs=None, extent=None, cellsize=None, dimensions=None):
+        self.__jvm = gps.get_spark_context()._gateway.jvm
+
+        if isinstance(layout, gps.LocalLayout):
+            if not extent:
+                raise ValueError("Must specify an extent when using LocalLayout")
+
+            if dimensions and not cellsize:
+                cellsize = ((extent.xmax - extent.xmin)/dimensions[0], (extent.ymax - extent.ymin)/dimensions[1])
+                dimensions = None
+
+            if cellsize and not dimensions:
+                tilewidth = layout.tile_cols * cellsize[0]
+                tileheight = layout.tile_rows * cellsize[1]
+                rows = ceil((extent.xmax - extent.xmin) / tilewidth)
+                cols = ceil((extent.ymax - extent.ymin) / tileheight)
+                extent = gps.Extent(extent.xmin, extent.ymax - rows * tileheight, extent.xmin + cols * tilewidth, extent.ymax)
+                tl = gps.TileLayout(cols, rows, layout.tile_cols, layout.tile_rows)
+            else:
+                raise ValueError("For LocalLayout, must specify exactly one: cellsize or dimension")
+        elif isinstance(layout, gps.GlobalLayout):
+            from pyproj import Proj, transform
+
+            if not crs:
+                raise ValueError("Must specify a crs when using GlobalLayout")
+
+            if isinstance(crs, int):
+                crs = "{}".format(crs)
+
+            gtcrs = self.__jvm.geopyspark.geotrellis.TileLayer.getCRS(crs).get()
+
+            if gtcrs.epsgCode().isDefined() and gtcrs.epsgCode().get() == 3857:
+                extent = gps.Extent(-20037508.342789244, -20037508.342789244, 20037508.342789244, 20037508.342789244)
+            elif gtcrs.epsgCode().isDefined() and gtcrs.epsgCode().get() == 4326:
+                extent = gps.Extent(-180.0, -89.99999, 179.99999, 89.99999)
+            else:
+                llex = gps.Extent(-180.0, -89.99999, 179.99999, 89.99999)
+                proj4str = gtcrs.toProj4String()
+                target = Proj(proj4str)
+                xmin, ymin = target(llex.xmin, llex.ymin)
+                xmax, ymax = target(llex.xmax, llex.ymax)
+                extent = gps.Extent(xmin, ymin, xmax, ymax)
+
+            if not layout.zoom:
```

Comment:
I think this can be moved up to where the `crs` input is checked. Since there's no reason to do the work of determining the `extent` if `zoom` isn't defined.